Improve inference for builtin oeprators.

notoraptor opened the pull request:
Extracted from #421 

@breuleux @abergeron

notoraptor committed:
Add inference for builtin oeprators.

notoraptor committed:
Extend inference to methods for binary boolean operators.

notoraptor committed:
Re-add test_bad to fix coverage.
Reformat code.

notoraptor commented:
Rebased and updated.

breuleux commented on myia/infer/inferrers.py:
```diff
@@ -82,22 +82,168 @@ def getattr_inferrer(node, args, unif):
+        a_type = yield Require(a_node)
+        b_type = yield Require(b_node)
+
+        if isinstance(a_type, data.GenericBase) or isinstance(
+            b_type, data.GenericBase
+        ):
+            # If there is any generic in operands, just unify them.
+            return autils.unify(a_type, b_type)[0]
```
This assumes that both operands have the same type (otherwise they would not be unifiable), which is not always true. For instance, it is untrue with `==`.

This is a delicate problem to solve and it might require to implement the inference subsystem that I mentioned a few weeks ago, where the inference of anything that involves generics is delayed as much as possible. The thing is, if `a_type` is a generic, trying to unify it with `b_type` is basically a guess, and we'd like to avoid ever having to guess.

To be clear I'm not asking for this feature in this PR, but tests that require this unification will have to be marked xfail.

notoraptor replied:
Ok. If I deactivate these lines, tests with `fact` will fail in `x * fact(x - 1)`. It seems it's because of `fact(x - 1)`, whose type is still Generic at this step.

I may just add a TODO here with your remarks, while keeping current code. Is it ok, or should I mark related tests as xfail in this PR ?

breuleux commented on myia/infer/inferrers.py:
```diff
@@ -82,22 +82,168 @@ def getattr_inferrer(node, args, unif):
+            new_node = node.graph.apply(
+                getattr(a_interface, bin_op), a_node, b_node
+            )
+        elif hasattr(b_interface, bin_rop):
```
Just adding a TODO to make sure we don't forget to adjust this when we have the capability (not yet).

```suggestion
        # TODO: In order to support user-defined graphs for bin_op, the case where
        # it returns NotImplemented will have to be handled.
        elif hasattr(b_interface, bin_rop):
```

notoraptor replied:
Ok!

breuleux commented on myia/infer/infnode.py:
```diff
@@ -80,6 +80,18 @@ def _type_to_abstract(el):
     return type_to_abstract(el)
 
 
+def _without_value(abstract: data.AbstractAtom):
```
Missing the decorator, also this is basically the `broaden` function that we had before.

```suggestion
@abstract_map.variant
def broaden(abstract: data.AbstractAtom):
```

notoraptor replied:
Ok!

breuleux commented on myia/infer/infnode.py:
```diff
@@ -80,6 +80,18 @@ def _type_to_abstract(el):
+
+    Useful to lookup node abstract in registered signatures.
+
+    NB: We must remove value track, not keeping it with ANYTHING value,
+    because abstract atom with only interface won't have same hash
+    as abstract atom with interface + ANYTHING value.
```
True, but that may need fixing...

notoraptor replied:
Ok, I add a TODO here.

breuleux commented on myia/infer/infnode.py:
```diff
@@ -80,6 +80,18 @@ def _type_to_abstract(el):
+    because abstract atom with only interface won't have same hash
+    as abstract atom with interface + ANYTHING value.
+    """
+    return data.AbstractAtom({"interface": abstract.tracks.interface})
```
You're removing all tracks that are not `interface`, not just `value`.

notoraptor replied:
Indeed, fixed, only value track removed.

breuleux commented on myia/infer/infnode.py:
```diff
@@ -134,12 +152,27 @@ def inference(node, args, unif):
+                f"No inference for node: {node}, "
+                f"signature: {list(inp_types)}"
+            )
+            return (yield from default(node, args, unif))
```
It may not be a bad idea to cache the result in `def_map`.

notoraptor replied:
Ok !

breuleux commented on tests/ir/test_print.py:
```diff
@@ -105,47 +105,47 @@ def test_format_exc():
-In add:clone(x::*int(), y::*float())
-85     return x + y
-              ^^^^^
+File ./tests/test_parser.py, line 2024
+In and_:clone(x~2::*int(), y::*float())
+2024     return f"value is {x}"
+         ^^^^^^^^^^^^^^^^^^^^^^
```
Why is it highlighting a line in `test_parser`?

notoraptor replied:
I still don't know, neither.

notoraptor committed:
Add a TODO to handle cases where bin op methods return NotImplemented.
Rename _without_value to broaden as a abstract_map variant.
Make broaden keep all tracks but value track.
In dispatch_inferences, cache inferred types from default inferred into def_map.

notoraptor committed:
Clear stack when calling enable_debug(), to prevent debug info passed from a previous graph parsing to a new one.

notoraptor commented:
Hi @breuleux , I add a commit to try to fix `format_exc` parsing. What do you think ?

breuleux commented on myia/utils/info.py:
```diff
@@ -40,6 +44,8 @@ def top(self):
 def enable_debug():
     """Enable debugging for a context."""
     tok = _debug.set(True)
+    # Clear stack each time we call enable_debug.
+    _stack.clear()
```
I think the problem happens because when an exception is raised in `with debug_inherit(...):`, the stack pop does not happen, so a better solution would be to add a `try: finally:` around the `yield` in `debug_inherit`, with the cleanup code in the `finally` block.

notoraptor replied:
OK, done. I still have 2 uncovered lines in `myia/ir/print`.

notoraptor committed:
Make sure to pop stack in debug_inherit() even if an exception occurs.

notoraptor committed:
Add another test format_exc_compact to fix coverage.

notoraptor commented:
Coverage fixed!
